src/undocommand.js

```diff
@@ -0,0 +1,192 @@
+/**
+ * @license Copyright (c) 2003-2016, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+'use strict';
+
+import Command from '../command/command.js';
+
+/**
+ * Undo command stores batches in itself and is able to and apply reverted versions of them on the document.
+ *
+ * undo.UndoCommand
+ */
+export default class UndoCommand extends Command {
+	/**
+	 * @see engine.command.Command
+	 * @param {engine.Editor} editor
+	 */
+	constructor( editor ) {
+		super( editor );
+
+		/**
+		 * Batches which are saved by the command. They can be reversed.
+		 *
+		 * @private
+		 * @member {Array.<engine.treeModel.Batch>} undo.UndoCommand#_batchStack
+		 */
+		this._batchStack = [];
+
+		/**
+		 * Stores the selection ranges for each batch and use them to recreate selection after undo.
+		 *
+		 * @private
+		 * @member {WeakMap.<engine.treeModel.Range>} undo.UndoCommand#_batchSelection
+		 */
+		this._batchSelection = new WeakMap();
```

Comment:
I do not want to keep the selection because it stores more information (i.e. attributes), it mixes `EmitterMixin`, it may grow and get some default events. We do not want to store `Selection` but values needed to recreate it. Keeping selection makes all the process more complicated. Instead saving just ranges + one flag you, additionally, have to:
1. Create a new `Selection` instance.
2. When restoring selection, get the `Selection` instance, get it's ranges.
This makes script more complicated and for what reason? I'd understand if we would restore the whole selection but you can't even say that "`Selection` may change so it's better to keep the instance" cause all we operate on are ranges and that's what we care about. So even if `Selection` changes, we will still just use `.setRanges()` method.